packages/kbn-custom-events/src/index.ts

```diff
@@ -0,0 +1,180 @@
+      const fullStoryChunkPromise = import('./fullstory');
+
+      // We need to call FS.identify synchronously after FullStory is initialized, so we must load the user upfront
+      const [{ initializeFullStory }, userId] = await Promise.all([
+        fullStoryChunkPromise,
+        userIdPromise,
+      ]);
+
+      const { fullStory, sha256 } = initializeFullStory({
+        basePath,
+        orgId: fullstoryOrgId,
+        packageInfo,
+      });
+
+      // This needs to be called syncronously to be sure that we populate the user ID soon enough to make sessions merging
+      // across domains work
+      if (userId) {
+        // Join the cloud org id and the user to create a truly unique user id.
+        // The hashing here is to keep it at clear as possible in our source code that we do not send literal user IDs
+        const hashedId = sha256(`${esOrgId}:${userId}`);
+        const kibanaVer = packageInfo.version;
+        // TODO: use semver instead
+        const parsedVer = (kibanaVer.indexOf('.') > -1 ? kibanaVer.split('.') : []).map((s) =>
+          parseInt(s, 10)
+        );
+        fullStory.identify(
+          hashedId,
+          this.formatContext({
+            esOrgId,
+            version: kibanaVer,
+            versionMajor: parsedVer[0] ?? -1,
+            versionMinor: parsedVer[1] ?? -1,
+            versionPatch: parsedVer[2] ?? -1,
+          })
+        );
+        this.fullStory = fullStory;
+      }
+    } catch (e) {
+      // eslint-disable-next-line no-console
+      console.error(`[custom events] Could not call initialize due to error: ${e.toString()}`, e);
+    }
+
+    return !!this.fullStory;
+  }
+
+  private getFullstoryType(value: FullstoryEventValue) {
+    // For arrays, make the decidion based on the first element
+    const v = isArray(value) ? value[0] : value;
+    switch (typeof v) {
+      case 'string':
+        return 'str';
+      case 'number':
+        return isInteger(value) ? 'int' : 'real';
+      case 'boolean':
+        return 'bool';
+      case 'object':
+        if (v instanceof Date) {
+          return 'date';
+        }
+    }
+  }
+
+  private getCurrentMemoryState() {
+    // Get performance information from the browser (non standard property
+    // @ts-expect-error 2339
+    const memory = window.performance.memory;
+    let memInfo = {};
+    if (memory) {
+      memInfo = {
+        jsHeapSizeLimit: memory.jsHeapSizeLimit,
+        jsHeapSizeTotal: memory.totalJSHeapSize,
+        jsHeapSizeUsed: memory.usedJSHeapSize,
+      };
+    }
+    return memInfo;
+  }
+
+  private formatContext(context: Record<string, FullstoryEventValue>) {
+    // format context keys as required for env vars, see docs: https://help.fullstory.com/hc/en-us/articles/360020623234
+    return mapKeys(context, (value, key) => {
+      const snakeCasedKey = snakeCase(key);
+      const type = this.getFullstoryType(value);
+      return [snakeCasedKey, type].join('_');
```

Comment:
Does it mean that we won't be able to correlate properties between different data storages unless all of them use `snake_case` formatting?

Reply:
This is a shctick with Fullstory. They enforce the format of field names.
In the current implementation (`master`) we just use the right format. Here I moved it to be an implementation detail of the integration.

Reply:
I'd love to _force_ devs to maintain a documented schema of the fields that they'll report. This will make documentation much easier, and we could even pre-validate the data, and even identify potential PII offenders (`boolean`s are not likely to include PII while `string`s are great candidates to audit) But on top of that, it will ensure that plugins don't _overwrite_ each other's fields or even common context data.